### DISCUSSION TITLE
More buttons, more flexibility

### DISCUSSION THREAD

**Discussion by probonopd** [probonopd/MiniDexed#310] [Announcements]
Since today's build, there is now more flexibility to configure the buttons for MiniDexed in custom ways. Instead of having just one button (the one built into the rotary encoder) you can now have multiple buttons, and for each button you can define which action gets triggered as you click, double-click, or long-press.

__PLEASE NOTE:__ `minidexed.ini` has changed. Please update yours accordingly when updating MiniDexed. Please see the sections "Buttons", "Timeouts", and "Rotary Encoder" at https://github.com/probonopd/MiniDexed/wiki/Files for details.
 
* __Default behavior changed__: Click to enter menus, long-press to leave menus, double-click to go to the home screen (configurable)
* Buttons can have click, double click and long press events
* Changed default config, added new options and renamed `EncoderPinSwitch` to `ButtonPinShortcut`

Thanks @diyelectromusic and @stevelittlefish who made this possible, and thanks to everyone who contributed with ideas, testing, and feedback!

Please post here what you are doing with the new possibilities. Which cool new button arrangements will you come up with in your custom device builds?

**Comment by antman49443** [comment]
Can you advise how to wire these buttons up? Or should I assume the hardware wiki will be updated shortly? Hope this works with a Pi1! Thanks!

**Reply by diyelectromusic** [reply]
They are all using internal pull-ups, so any button just has to short to ground for it to register.  Use whatever pins make sense, but naturally you'll need to watch out for GPIO used elsewhere...

It should work fine with a Pi 1 (again as long as you have enough pins :))

**Reply by antman49443** [reply]
OK, this works pretty good! In my opinion, much better than with just the encoder. On my Pi 1, rev2, I'm using GPIO 27 for PREV, 18 for NEXT, 7 for BACK, 8 for Home , and 11 stays SELECT. I'm not using a separate DAC, so that probably limits buttons for the P1.

**Comment by Doug2233** [comment]
Does the ‘long press back’ operation interfere with the option to ‘press and hold then turn to select the same option’ for other TGs?  I find this option useful when setting channel numbers.  Thanks

**Reply by stevelittlefish** [reply]
No - as long as you turn the encoder before the long press timeout, it will cancel the long press and do the "press and turn" shortcut instead.

**Reply by Doug2233** [reply]
Thank you!

**Comment by kyohejr** [comment]
Would it be possible to have a second optional rotary encoder to make it easier to switch between voices and bank?  For example, if the primary encoder is in the voice menu, the second encoder could cycle between banks. The function of the second encoder would be dependent on the mode for the primary encoder.

**Reply by probonopd** [reply]
Wouldn't hat make things a) unneccesarily complicated and b) unneccesarily expensive?

**Reply by kyohejr** [reply]
I thought the encoders were relatively inexpensive.  Though I saw another suggestion to automatically switch to the next bank when you cycle up from voice 32. Perhaps this is a better solution. Probably a lot simpler.  Personally for me the biggest hassle with this is remembering which bank has the voice I’m trying to find.

**Reply by dlbogdan** [reply]
I'm thinking of branching MiniDexed to be able to use multiple rotary encoders in conjunction with some spi or i2c port expanders. I'm very new to music, in fact, I just want to build an awesome piano for my son to play with, but also have a little fun in the process. In the end I intend to ditch the menu, and replace that with just status indicators on a larger screen, and have a physical knob for mostly everything configurable.

**Reply by beeltink** [reply]
I'd be interested in preset-buttons. Would that be hard?

**Reply by diyelectromusic** [reply]
Do you mean a whole range of buttons to choose the voices?  If so then I think that is better achieved with a separate unit sending MIDI Program Change messages rather than trying to do it all in the MiniDexed code. 

I did something like that with an Arduino here: https://github.com/probonopd/MiniDexed/discussions/236

Kevin

**Reply by Banana71** [reply]
Unfortunately it is not yet possible to switch Performances via MIDI ProgrammChange.  So you can only change in the TG's Voices on the selected Midi channel

**Reply by diyelectromusic** [reply]
Is this something we'd want?  An option to make PC messages select performances rather than voices?  It would need some thinking about (i.e. which MIDI channels to be active on from which TGs, etc) but that probably isn't particularly difficult I think to make happen?

I think it would have to be a config setting to make PC messages either voices or performances though. I don't see how you could have both... unless we had a special virtual "bank" that mapped onto performances... but that is getting a little complicated!

Kevin

**Reply by Banana71** [reply]
The TX802 has a parameter "Performance Select" 
- "off" no program change
-  "1 -16" selects the midi channel on which program changes for performance are received.  
- "normal" program changes affect the TG's

**Reply by Kirtai** [reply]
> I think it would have to be a config setting to make PC messages either voices or performances though. I don't see how you could have both... unless we had a special virtual "bank" that mapped onto performances... but that is getting a little complicated!

Isn't that what banks are for?

**Reply by diyelectromusic** [reply]
Well banks are (currently) collections of voices not collections of performances...

**Reply by Kirtai** [reply]
Yes, but it should be possible to assign one or more banks to performances. If that's the way the devs want to go.

**Reply by probonopd** [reply]
@Kirtai please see what @diyelectromusic is proposing in https://github.com/probonopd/MiniDexed/pull/500, and ideally give the test build a try.
